File: src-tauri/src/commands/software.rs
/**
 * software.rs - 软件包 CRUD 和设置命令
 */
use log::{debug, info};
use tauri::State;

use crate::models::*;
use crate::AppState;

/// 获取所有软件包列表
#[tauri::command]
pub async fn list_software(state: State<'_, AppState>) -> Result<Vec<SoftwareInfo>, String> {
    debug!("正在获取所有软件包列表");
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let result = db.get_all_software().map_err(|e| e.to_string())?;
    info!("已列出 {} 个软件包", result.len());
    Ok(result)
}

/// 获取软件包列表展示数据（含 AUR + Upstream 信息）
#[tauri::command]
pub async fn list_software_view(state: State<'_, AppState>) -> Result<Vec<SoftwareListEntry>, String> {
    debug!("正在获取软件包视图列表");
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let result = db.get_software_list_entries().map_err(|e| e.to_string())?;
    info!("已列出 {} 个软件包视图条目", result.len());
    Ok(result)
}

/// 根据包名获取单个软件包信息
#[tauri::command]
pub async fn get_software(
    state: State<'_, AppState>,
    pkgname: String,
) -> Result<Option<SoftwareInfo>, String> {
    debug!("正在获取软件包: {}", pkgname);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.get_software_by_name(&pkgname).map_err(|e| e.to_string())
}

/// 根据包名获取软件包完整详情（含 AUR + 上游信息）
#[tauri::command]
pub async fn get_software_detail(
    state: State<'_, AppState>,
    pkgname: String,
) -> Result<Option<SoftwareDetail>, String> {
    debug!("正在获取软件包详情: {}", pkgname);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.get_software_detail_by_name(&pkgname)
        .map_err(|e| e.to_string())
}

/// 获取上一个和下一个软件包名称（用于导航）
#[tauri::command]
pub async fn get_prev_next_software(
    state: State<'_, AppState>,
    pkgname: String,
) -> Result<(Option<String>, Option<String>), String> {
    debug!("正在获取上一个/下一个软件包: {}", pkgname);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.get_prev_next_software(&pkgname)
        .map_err(|e| e.to_string())
}

/// 搜索软件包
#[tauri::command]
pub async fn search_software(
    state: State<'_, AppState>,
    keyword: String,
) -> Result<Vec<SoftwareInfo>, String> {
    debug!("正在搜索软件包: {}", keyword);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let result = db.search_software(&keyword).map_err(|e| e.to_string())?;
    info!("搜索 '{}' 找到 {} 个结果", keyword, result.len());
    Ok(result)
}

/// 添加新的软件包
#[tauri::command]
pub async fn add_software(
    state: State<'_, AppState>,
    pkgname: String,
    upstream_url: Option<String>,
    package_type: i32,
    checker_type: i32,
    check_test_versions: bool,
    check_binary_files: bool,
    auto_check_enabled: bool,
    license_id: Option<i64>,
    language_id: Option<i64>,
) -> Result<i64, String> {
    info!("正在添加软件包: {}", pkgname);
    let sw = SoftwareInfo {
        software_id: None,
        pkgname: pkgname.clone(),
        upstream_url,
        package_type_id: PackageType::from_id(package_type),
        checker_type_id: CheckerType::from_id(checker_type),
        is_outdated: false,
        check_test_versions,
        check_binary_files,
        auto_check_enabled,
        license_id,
        language_id,
    };
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let id = db.insert_software(&sw).map_err(|e| e.to_string())?;
    info!("已添加软件包 '{}'，ID: {}", pkgname, id);
    Ok(id)
}

/// 更新软件包信息（编辑）
#[tauri::command]
pub async fn update_software(
    state: State<'_, AppState>,
    software_id: i64,
    pkgname: String,
    upstream_url: Option<String>,
    package_type: i32,
    checker_type: i32,
    is_outdated: bool,
    check_test_versions: bool,
    check_binary_files: bool,
    auto_check_enabled: bool,
    license_id: Option<i64>,
    language_id: Option<i64>,
) -> Result<(), String> {
    info!("正在更新软件包 {}: {}", software_id, pkgname);
    let sw = SoftwareInfo {
        software_id: Some(software_id),
        pkgname: pkgname.clone(),
        upstream_url,
        package_type_id: PackageType::from_id(package_type),
        checker_type_id: CheckerType::from_id(checker_type),
        is_outdated,
        check_test_versions,
        check_binary_files,
        auto_check_enabled,
        license_id,
        language_id,
    };
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.upsert_software(&sw).map_err(|e| e.to_string())
}

/// 设置软件包的 License
#[tauri::command]
pub async fn set_software_license(
    state: State<'_, AppState>,
    software_id: i64,
    license_id: Option<i64>,
) -> Result<(), String> {
    info!("正在设置软件包 {} 的 License 为 {:?}", software_id, license_id);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.update_software_license(software_id, license_id)
        .map_err(|e| e.to_string())
}

/// 设置软件包的编程语言
#[tauri::command]
pub async fn set_software_language(
    state: State<'_, AppState>,
    software_id: i64,
    language_id: Option<i64>,
) -> Result<(), String> {
    info!("正在设置软件包 {} 的编程语言为 {:?}", software_id, language_id);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.update_software_language(software_id, language_id)
        .map_err(|e| e.to_string())
}

/// 删除单个软件包
#[tauri::command]
pub async fn delete_software(state: State<'_, AppState>, software_id: i64) -> Result<(), String> {
    info!("正在删除软件包，ID: {}", software_id);
    let db = state.db.lock().map_err(|e| e.to_string())?;
    db.delete_software(software_id).map_err(|e| e.to_string())
}

/// 批量删除软件包
#[tauri::command]
pub async fn batch_delete_software(
    state: State<'_, AppState>,
    ids: Vec<i64>,
) -> Result<i64, String> {
    info!("正在批量删除 {} 个软件包", ids.len());
    let db = state.db.lock().map_err(|e| e.to_string())?;
    let mut count = 0i64;
    for id in &ids {
        if db.delete_software(*id).is_ok() {
            count += 1;
        }
    }
    info!("已删除 {} 个软件包", count);
    Ok(count)
}

File: src/api/software.ts
/**
 * software.ts - 软件包领域 API 封装
 *
 * 功能：
 * - 集中封装「软件包 CRUD / AUR 同步 / 上游版本检查 / URL 校验」相关的 Tauri 命令调用
 * - 组件与 composable 统一通过本模块访问后端，避免 invoke 命令字符串散落各处
 *
 * 设计：每个函数对应一个后端 Tauri 命令，参数与返回类型与原有 invoke 调用保持一致。
 */

import { invoke } from "@tauri-apps/api/core";
import type {
  SoftwareInfo,
  SoftwareDetail,
  SoftwareListEntry,
  ValidateResult,
} from "@/types";

/** 新增软件包入参（packageType / checkerType 在前端表单中以 number 建模，与后端 i32/i64 对应） */
export interface AddSoftwareInput {
  pkgname: string;
  upstreamUrl: string | null;
  packageType: number;
  checkerType: number;
  checkTestVersions: boolean;
  checkBinaryFiles: boolean;
  autoCheckEnabled: boolean;
  languageIds: number[];
  versionExtractRegex: string | null;
}

/** 更新软件包入参（在新增入参基础上附带 ID 与过期标记） */
export interface UpdateSoftwareInput extends AddSoftwareInput {
  softwareId: number;
  isOutdated: boolean;
}

/** 获取相邻软件包（上一包 / 下一包 pkgname，无则 null） */
export async function getPrevNextSoftware(
  pkgname: string,
): Promise<[string | null, string | null]> {
  return await invoke<[string | null, string | null]>("get_prev_next_software", { pkgname });
}

/** 按 pkgname 获取软件包完整详情 */
export async function getSoftwareDetail(pkgname: string): Promise<SoftwareDetail | null> {
  return await invoke<SoftwareDetail | null>("get_software_detail", { pkgname });
}

/** 按 pkgname 获取列表条目（轻量视图） */
export async function getSoftwareListEntry(pkgname: string): Promise<SoftwareListEntry | null> {
  return await invoke<SoftwareListEntry | null>("get_software_list_entry", { pkgname });
}

/** 列出全部软件包（完整信息） */
export async function listSoftware(): Promise<SoftwareInfo[]> {
  return await invoke<SoftwareInfo[]>("list_software");
}

/** 列出软件包列表视图（分页/展示用轻量结构） */
export async function listSoftwareView(): Promise<SoftwareListEntry[]> {
  return await invoke<SoftwareListEntry[]>("list_software_view");
}

/** 新增软件包，返回新记录 software_id */
export async function addSoftware(input: AddSoftwareInput): Promise<number> {
  return await invoke<number>("add_software", { ...input });
}

/** 更新软件包 */
export async function updateSoftware(input: UpdateSoftwareInput): Promise<void> {
  await invoke("update_software", { ...input });
}

/** 删除单个软件包 */
export async function deleteSoftware(softwareId: number): Promise<void> {
  await invoke("delete_software", { softwareId });
}

/** 批量删除软件包 */
export async function batchDeleteSoftware(pkgnameList: string[]): Promise<void> {
  await invoke("batch_delete_software", { pkgnameList });
}

/** 设置软件包关联的 License（单个 AUR license 名称，null 表示清除） */
export async function setSoftwareLicense(softwareId: number, licenseId: string | null): Promise<void> {
  await invoke("set_software_license", { softwareId, licenseId });
}

/** 更新选中包的 AUR 信息（pkgnameList 为 null 表示全部） */
export async function updateAurInfo(pkgnameList: string[] | null): Promise<number> {
  return await invoke<number>("update_aur_info", { pkgnameList });
}

/** 从 AUR 全量同步软件包信息 */
export async function syncFromAur(): Promise<void> {
  await invoke("sync_from_aur");
}

/** 从 PKGBUILD 同步（pkgname 为 null 表示全部） */
export async function syncFromPkgbuild(pkgname: string | null): Promise<void> {
  await invoke("sync_from_pkgbuild", { pkgname });
}

/** 检查单个软件包的上游版本，返回最新上游版本号 */
export async function checkUpstreamVersion(pkgname: string): Promise<string> {
  return await invoke<string>("check_upstream_version", { pkgname });
}

/** 检查选中软件包的上游版本，返回 [(pkgname, version)] */
export async function checkSelectedUpstream(pkgnameList: string[]): Promise<[string, string][]> {
  return await invoke<[string, string][]>("check_selected_upstream", { pkgnameList });
}

/** 检查全部软件包的上游版本，返回 [(pkgname, version)] */
export async function checkAllUpstream(): Promise<[string, string][]> {
  return await invoke<[string, string][]>("check_all_upstream");
}

/** 校验上游 URL 可达性 */
export async function validateUpstreamUrls(pkgnameList: string[] | null): Promise<ValidateResult[]> {
  return await invoke<ValidateResult[]>("validate_upstream_urls", { pkgnameList });
}
